website_portal_medical_insurance_us/controllers/main.py

```diff
@@ -0,0 +1,69 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016 LasLabs Inc.
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
+
+from odoo import fields, http
+from odoo.http import request
+
+from odoo.addons.website_portal_medical_patient.controllers.main import (
+    WebsiteMedical
+)
+
+
+class WebsiteMedical(WebsiteMedical):
+
+    @http.route(
+        ['/my/medical', '/medical'],
+        type='http',
+        auth="user",
+        website=True,
+    )
+    def my_medical(self, **kw):
+        """ Add insurances to medical account page """
+        response = super(WebsiteMedical, self).my_medical()
+        active_plans = request.env['medical.insurance.plan'].\
+            get_related_patient_medical_plans()
+        response.qcontext.update({
+            'insurance_plans': active_plans.sudo(),
+        })
+        return response
+
+    @http.route(
+        ['/medical/insurance/plans'],
+        type='http',
+        auth='user',
+        website=True,
+        methods=['GET'],
+    )
+    def medical_insurance(self):
+        active_plans = request.env['medical.insurance.plan'].\
+            get_related_patient_medical_plans()
+        values = {
+            'user': request.env.user,
+            'insurance_plans': active_plans,
+        }
+        return request.render(
+            'website_portal_medical_insurance_us.insurances',
+            values,
+        )
+
+    @http.route(
+        ['/medical/insurance/plans/<model("medical.insurance.plan"):plan>'],
+        type='http',
+        auth='user',
+        website=True,
+        methods=['GET'],
+    )
+    def insurance_plan_show(self, plan=None, redirect=None, **kwargs):
```

Comment:
@lasley for this method. When the user wants to create a new insurance plan, the 'ADD' link will pass in `/medical/insurance/plans/0`. Which will show a 404 page.

Before, you instead had the `<int:id>` or something like that instead of the `<model...`. in the url. With your version we could browse and get an empty record and the form would display. When using `<model . . .` I don't think that's possible as it seems I get a 404 before the method logic even runs. I think awhile ago you were mentioning `<model ...` does validations to ensure the record exists.

With that in mind, to stay consistent, should we:
* Have all urls use the `<int:id>` to account for creating new recs instead of `<model...` or have something like:
* `/medical/insurance/plans/new` for creating new recs. That way the logic for showing forms for existing records and for showing a form to create a new one are separated.

Reply:
Ah dang sorry it doesn't show the method's url in preview. The URL for this method is:
```python
    @http.route(
        ['/medical/insurance/plans/<model("medical.insurance.plan"):plan>'],
        type='http',
        auth='user',
        website=True,
        methods=['GET'],
    )
    def insurance_plan_show(self, plan=None, redirect=None, **kwargs):
```

Reply:
No need for a `/new`, just listen on `/medical/insurance/plans` for `POST` & act on that instead?

Reply:
Also it should all be singular `/plan` not `plans`. My bad there

Reply:
ah I see. Those should be modelled after the model name, not pluralized. I had changed to plans to stay consistent with patients. I'll need to edit the patient portals then too back to patient and not patients.

For the case of POST though, that's covered by your website_form method in website_portal_medical.

Perhaps I'm misunderstanding, but what I'm trying to get at is let's say the user is at `/medical/insurance/plan`, displaying all the plans. There is a link on that page that says `ADD` with an href to `/medical/insurance/plan/0`. This is to display a form to create a new plan.

Your code is as follows (at the bottom), which allows the rendering of the view even if no record exists. This allows for the display of a form when a user wants to create a new plan.

However if I change to ```['/medical/insurance/plan/<model("medical.insurance.plan"):plan>']```, I get a 404.

So I'm just trying to figure out what would be the best way to display a form for a record that does not exist in the database yet.

Should we have a `/medical/insurance/plan/new` uri specified as `GET`? Or should the logic be as what you have done?

Or perhaps what you specified in your comment using post. That doesn't really make sense to me to use post though.

```python
    @http.route(
        ['/medical/insurance/plan/<int:plan_id>'],
        type='http',
        auth='user',
        website=True,
        methods=['GET'],
    )
    def insurance_plan_show(self, plan_id=None, redirect=None, **kwargs):
        values = {
            'error': {},
            'error_message': [],
            'success_page': kwargs.get('success_page', '/my/medical')
        }
        Plans = http.request.env['medical.insurance.plan']
        if plan_id:
            plan = Plans.browse(plan_id)
        else:
            plan = Plans.browse()
        values.update({
            'insurance_plan': plan and plan.sudo(),
        })
        return request.render(
            'website_portal_medical_insurance_us.insurance_plan', values,
        )
```

Reply:
Reason I'm asking is that medical patient portal is following what I have done, using model in the uri. 

Upon looking here, it made me realize you can't create a patient from the frontend. So just wanted to confirm to either change to what you are doing for insurance plans, or if another idea is better.

Reply:
`/plans` would display all plans, `/plan/x` would would display plan ID x. `/plan` should display a blank form. We should align this everywhere

Reply:
Awesome that clears things up. Sorry for the long comments : \